File: join.js
var i, k, t, d1, d2, d2_hash, result_hash, result;

k = 1000 * 1000;
k = 100 * 1000;

console.log('K =', k);

t = new Date();
var d1 = [];
for (i = 0; i < 10 * k; i++) {
    d1.push([1, i, 10]);
}
console.log('%s\t%s', (new Date() - t) / 1000, 'd1 list');

t = new Date();
d2 = [];
for (i = 2 * k; i < 3 * k; i++) {
    d2.push([1, i, 5]);
}
console.log('%s\t%s', (new Date() - t) / 1000, 'd2 list');

console.log('\n--- start join');
tjoin = new Date();

t = new Date();
d2_hash = {};
d2.forEach(function (item) {
    d2_hash[item[0] + '-' + item[1]] = item[2];
});
console.log('%s\t%s', (new Date() - t) / 1000, 'd2 hash');

t = new Date();
result_hash = {};
d1.forEach(function (item) {
    var k = item[0] + '-' + item[1];
    result_hash[k] = [item[2], d2_hash[k]];
});
console.log('%s\t%s', (new Date() - t) / 1000, 'd1 part of result hash');

t = new Date();
d2.forEach(function (item) {
    var k = item[0] + '-' + item[1];
    if (!(k in result_hash)) {
        result_hash[k] = [undefined, item[2]];
    }
});
console.log('%s\t%s', (new Date() - t) / 1000, 'add d2 part of result hash');

console.log('--- full join time:', (new Date() - tjoin) / 1000, '\n');

t = new Date();
result = [];
for (k in result_hash) {
    var v = result_hash[k];
    result.push([k, v[0], v[1]]);
}
console.log('%s\t%s', (new Date() - t) / 1000, 'format result_hast to table');

for (i = 0; i < 10 && i < result_hash.length; i++) {
    console.log(result[i]);
}

File: join.go
package main

import (
	"fmt"
	"time"
)

func main() {
	k := 1000 * 1000
	k = 100 * 1000

	fmt.Println("K = ", k)

	t := time.Now()
	d1 := make([][]int, 10*k)
	for i := range d1 {
		d1[i] = []int{1, i, 10}
	}
	fmt.Printf("%.8f\t d1 list\n", time.Since(t).Seconds())

	t = time.Now()
	d2 := make([][]int, k)
	for i := range d2 {
		d2[i] = []int{1, i + 2*k, 5}
	}
	fmt.Printf("%.8f\t d2 list\n", time.Since(t).Seconds())

	tjoin := time.Now()
	fmt.Println("\n--- start join")

	t = time.Now()
	d2_hash := make(map[string]int, len(d2))
	for i := range d2 {
		k := fmt.Sprintf("%d-%d", d2[i][0], d2[i][1])
		d2_hash[k] = d2[i][2]
	}
	fmt.Printf("%.8f\t d2 hash\n", time.Since(t).Seconds())

	t = time.Now()
	result_hash := map[string]string{}
	for i := range d1 {
		k := fmt.Sprintf("%d-%d", d1[i][0], d1[i][1])
		d2_v, found := d2_hash[k]
		if found {
			result_hash[k] = fmt.Sprintf("%d-%d", d1[i][2], d2_v)
		} else {
			result_hash[k] = fmt.Sprintf("%d-%d", d1[i][2], 0)
		}
	}
	fmt.Printf("%.8f\t d1 part to result hash\n", time.Since(t).Seconds())

	t = time.Now()
	for i := range d2 {
		k := fmt.Sprintf("%d-%d", d2[i][0], d2[i][1])
		_, found := result_hash[k]
		if !found {
			result_hash[k] = fmt.Sprintf("%d-%d", 0, d2[i][2])
		}
	}
	fmt.Printf("%.8f\t add d2 part to result hash\n", time.Since(t).Seconds())

	fmt.Printf("--- full join time: %.8f\n\n", time.Since(tjoin).Seconds())

	t = time.Now()
	result := make([][]string, len(result_hash))
	i := 0
	for k, v := range result_hash {
		result[i] = []string{k, v}
		i++
	}
	fmt.Printf("%.8f\t format result_hast to table\n\n", time.Since(t).Seconds())

	fmt.Println(result[:10])

}
